### PR TITLE
Compile Action Cable assets to JavaScript

### PR DESCRIPTION
**Pull request by maclover7**
TODO:
- figure out how to make it so CoffeeScript version will only be required if `coffee-rails` is in the bundle. See `engine.rb` for some preliminary work on this.

cc @javan 
r? @matthewd

### CODE REVIEW

File: actioncable/Rakefile [hunk 1]
```diff
@@ -11,3 +13,40 @@ Rake::TestTask.new do |t|
   t.verbose = true
   t.ruby_opts = ["--dev"] if defined?(JRUBY_VERSION)
 end
+
+namespace :assets do
+  task :precompile do
+    ## Check for coffee executable
+    require 'mkmf'
+    unless find_executable('coffee')
+      raise "In order to compile Action Cable's assets, you must have the "\
```
Comment (rafaelfranca) [inline actioncable/Rakefile]: We have sprockets, why don't use it?

**Comment by rafaelfranca** [comment]
> figure out how to make it so CoffeeScript version will only be required if coffee-rails is in the bundle.

Why? Can't people just use the javascript version without any problem?

**Comment by rafaelfranca** [comment]
The JS files don't need to be in the repository. We can generate them when building the gem.

File: actioncable/Rakefile [hunk 2]
```diff
@@ -11,3 +13,40 @@ Rake::TestTask.new do |t|
...
+      ] + lines
+
+      output = File.new("#{ASSETS_ROOT}/action_cable_js/consumer.js", "w")
+      lines.each { |line| output.write line }
+      output.close
+
+      ## Output results
+      puts 'All done!'
+      puts '**Please note:** Due to action_cable.coffee.erb having a .erb'\
+        ' extension, you will need to manually compile this.'
```
Comment (matthewd) [inline actioncable/Rakefile]: We can fix this by pushing `ActionCable.INTERNAL` into a separate file, and just making it `.js.erb`

File: actioncable/Rakefile [hunk 3]
```diff
@@ -11,3 +13,40 @@ Rake::TestTask.new do |t|
...
+      ## Add requires for consumer.js
+      f = File.open("#{ASSETS_ROOT}/action_cable_js/consumer.js", "r+")
+      lines = f.readlines
+      f.close
+
+      lines = [
+        "//= require action_cable/connection\n",
+        "//= require action_cable/connection_monitor\n",
+        "//= require action_cable/subscriptions\n",
+        "//= require action_cable/subscription\n"
```
Comment (matthewd) [inline actioncable/Rakefile]: If we compile via sprockets, as @rafaelfranca suggested, then we should be able to skip these complications, and just produce a single combined JS file (plus a require for the 'internal' thing, mentioned below).

**Comment by javan** [comment]
What's the motivation here? I'm not suggesting it's a bad idea, but there's no explanation of why in this PR. New Rails apps ship with CoffeeScript enabled by default and include Turbolinks, which is also uncompiled.

**Comment by matthewd** [comment]
@javan the motivation is that it's a release blocker; Rails does not currently force people to install coffeescript, and changing that would be a bad thing.

**Comment by vipulnsward** [comment]
Since all the other official JS gems(jquery-rails/ jquery-ujs) come with JS instead of Coffeescript, would it be good idea to move from CS to JS?

This is going to create issues that might be fixed on JS vs CoffeeScript issues.

**Comment by javan** [comment]
By that measure, Turbolinks should be a release blocker.

**Comment by javan** [comment]
Perhaps we should come up with a general purpose solution for compiling to .js instead of building it in to Action Cable.

**Comment by javan** [comment]
> The JS files don't need to be in the repository. We can generate them when building the gem.

Generate them and put them where? Also, I think we should preserve the ability to point your Gemfile to master and use the latest assets.

**Comment by rafaelfranca** [comment]
> By that measure, Turbolinks should be a release blocker.

No. Turbolinks is not a dependecy of `rails` gem so you can just skip both turbolinks and coffeescript and have an application without coffescript but your can't skip actioncable.

I think we only need to generate the JS files when building the gem and if you want to point to master you will need coffeescript.

**Comment by javan** [comment]
Oh, my misunderstanding. I thought that `actioncable` was an optional gem in an app's Gemfile.

**Comment by rafaelfranca** [comment]
No problem.

**Comment by kaspth** [comment]
Why do we need to move the assets into a `rails_assets` folder? Won't the existing `assets` folder work just as well?

**Comment by maclover7** [comment]
This is so that way we manually load in either the CoffeeScript or the equivalent compiled JavaScript. If everything's in `assets`, then we won't have control over which version is enabled for end users.

**Comment by kaspth** [comment]
Ah, do engines automatically include `assets/javascripts` in to the main application? Otherwise I don't see how we don't have control over the assets. It feels like there ought to be a better way to accomplish this.

Especially like @rafaelfranca's suggestion of adding the vanilla JS assets to the shipped bundle. Perhaps  then we could even replace the `assets/javascripts` contents with the compiled assets.

**Comment by javan** [comment]
I've come up with an alternative to this change that:

1. Uses Sprockets
2. Doesn't require rewriting / modifying the original source files

https://github.com/rails/rails/pull/23365

Let me know what you think.

**Comment by maclover7** [comment]
Closing in favor of #23369.

**Pull request closed without merging**